common/ferrostar/src/navigation_controller/step_advance/mod.rs

```diff
@@ -23,18 +23,49 @@ pub struct StepAdvanceResult {
     /// The next iteration of the step advance condition.
     ///
     /// This is what the navigation controller passes to the next instance of [`NavState`](super::NavState) on the completion of
-    /// an update (e.g. a user location update). Usually, this value is one of the following:
+    /// an update (e.g. a user location update). Use the helper methods on [`StepAdvanceResult`] to ensure proper state handling:
     ///
-    /// 1. When `should_advance` is true, this should typically be a clean/new instance of the condition.
-    /// 2. When the condition is not advancing, but the condition maintains no state, this should be a
-    ///    clean/new instance of the condition.
-    /// 3. When the condition is not advancing and maintains state, this should be a new
-    ///    instance including the current state of the condition. See [`DistanceEntryAndExitCondition`]
+    /// - [`StepAdvanceResult::next()`] - Unified method that takes a boolean and handles state appropriately
+    /// - [`StepAdvanceResult::advance()`] - Creates an advancing result with automatically reset state
+    /// - [`StepAdvanceResult::no_advance()`] - Creates a non-advancing result preserving current state
     ///
-    /// IMPORTANT! If the condition advances. This **must** be the clean/default state.
+    /// The trait method [`StepAdvanceCondition::new_instance()`] ensures that composite conditions
+    /// (Or/And) properly reset all nested conditions when any condition triggers advancement.
+    ///
+    /// **CRITICAL**: When advancing, this must be a clean/reset state to prevent state leakage between steps.
     pub next_iteration: Arc<dyn StepAdvanceCondition>,
 }
 
+impl StepAdvanceResult {
+    /// Creates a step advance result that does not advance to the next step.
+    /// Uses the provided next_iteration as-is for preserving stateful progress.
+    pub fn no_advance(next_iteration: Arc<dyn StepAdvanceCondition>) -> Self {
+        Self {
+            should_advance: false,
+            next_iteration,
+        }
+    }
+
+    /// Creates a step advance result that advances to the next step.
+    /// Automatically creates a fresh instance of the condition to ensure proper state isolation.
+    pub fn advance(condition: &dyn StepAdvanceCondition) -> Self {
+        Self {
+            should_advance: true,
+            next_iteration: condition.new_instance(),
+        }
+    }
+
+    /// Creates a step advance result based on the should_advance boolean.
+    /// If advancing, creates a fresh instance. If not advancing, uses the provided next_iteration.
+    pub fn next(should_advance: bool, condition: &dyn StepAdvanceCondition) -> Self {
```

Comment:
Can we make the API simpler here? It sems like, based on the descriptions, this is completely covered by `advance` and `no_advance`, and it's just a more confusing interface.

It also doesn't do what the docstring says; it creates the new condition instance only if we stay put (`no_advance`), which seems strange.